designs/0004-sparse-matrices.md

```diff
@@ -0,0 +1,387 @@
+- Reverse-mode derivative of a Cholesky decomposition of a sparse matrix.
+  - This is the same algorithm that was initially used in the dense case before Rob implemented Iain Murray’s blocked version. But it will need to be implemented in a “data structure” aware way
+- Specialization for Poisson, binomial and maybe gamma likelihoods.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Most of the below comes from the [functional-spec](https://github.com/stan-dev/stan/wiki/Functional-Spec:-Sparse-Matrix-Data-Types) for sparse matrices.
+
+A sparse matrix/vector differs from a standard matrix/vector only in the underlying representation of the data.
+
+Below we go over how sparse matrices can be constructed and operated on within each stan block.
+
+
+## Data
+
+Sparse matrix types in the Stan language can be constructed in the data block via Coordinate List Notation using the row and column sizes, non-empty row/column indices, and values for those index positions.
+
+```stan
+data {
+int N; // Rows
+int M; // Cols
+int K; // number non-empty
+int nonzero_row_index[K]; // Non-empty row positions
+int nonzero_col_index[K]; // Non-empty col positions
+vector[N] vals; // Values in each position
+// Direct way
+sparse_matrix[N, M, nonzero_row_index, nonzero_col_index, val] A
+// Can we do this?
+sparse_matrix[N, M, nonzero_row_index, nonzero_col_index] B;
+}
+```
+
+Sparse vectors are the same only with a single size on a single index array.
+
+```stan
+int N; // Rows
+int K; // number non-empty
+int nonzero_row_index[K]; // Non-empty row positions
+vector[N] vals; // Values in each position
+// Direct way
+sparse_matrix[N, nonzero_row_index, val] A
+// Can we do this?
+sparse_vector[N, nonzero_row_index] B;
+```
+
+The above sparse matrix example with values
+
+```stan
+N = 5
+M = 5
+K = 7
+// Column major order
+nonzero_row_index[K] = [2, 3, 1, 3, 5, 3, 2, 5]
+nonzero_col_index[K] = [1, 1, 2, 2, 3, 4, 5, 5]
+val[K] = [22, 7, 3, 5, 14, 1, 17, 8]
+```
+
+Would have the dense form of
+
+| row/col | 1  | 2 | 3  | 4 | 5  |
+|---------|----|---|----|---|----|
+| 1       | 0  | 3 | 0  | 0 | 0  |
+| 2       | 22 | 0 | 0  | 0 | 17 |
+| 3       | 7  | 5 | 0  | 1 | 0  |
+| 4       | 0  | 0 | 0  | 0 | 0  |
+| 5       | 0  | 0 | 14 | 0 | 8  |
+
+## Transformed Data
+
+Sparse matrices in this block can be defined dynamically and declared such as
+
+```stan
+transformed data {
+// Could construct here as well
+sparse_matrix[N, M] A = to_sparse_matrix(N, M, nonzero_row_index, nonzero_col_index, vals);
+
+// Linear Algebra is cool
+sparse_matrix[N, N] C = A * A';
+// For data this is fine
+C[10, 10] = 100.0;
+}
+```
+
+The assignment operation `C[10, 10] = 100.0;` works fine with Eigen as the implementation leaves room in the arrays for quick insertions.
+
+## Parameters, Transformed Parameters, and Generated Quantities
+
+Parameters be defined as above for data or deduced from the output of other functions.
```

Comment:
We don't have type inference yet in Stan - I think so far declarations will all need to be of the same form... not sure. Might want to make an open item of that in the design doc.

Reply:
> I think so far declarations will all need to be of the same form

One thing I'm a bit worried by is that users would then have to know the non-zero elements of the Cholesky or any sparse matrix output.

When I'm thinking about this, we need to specify the nonzero rows / columns when we are writing data into a sparse matrix via triplets. But the Eigen declaration for sparse matrices is really just `Eigen::SparseMatrix<T>`, so it doesn't really need to know the nonzero elements. So doing something like 

```
Eigen::SparseMatrix<double> A = inv_solver.solve(B);
```
Will not need to know the non-zero elements.

Does it make sense what I'm getting at here? i.e. that specifying the non-zero elements is much more for Eigen matrix initialization than it is for the math library.

Reply:
Yeah, that makes sense. It's very good that code generation can support not knowing the sparsity ahead of time for Eigen in the cases where it is immediately assigned to something with Eigen sparse structure. It makes it a little more difficult for the compiler to check that this is true, though probably still mostly possible. We'd need to basically add sparsity to our type system somehow, possibly just as two new Eigen types, SparseMatrix and SparseVector (there are no SparseRowVectors that we care about, right?). And then we can annotate all of the Stan Lib functions with additional sparsity type signatures so we know when a return value will be sparse given the input types. And then the compiler can check that we don't need the sparsity on the declaration. 

That brings up another awkward point - do we need to check the sparsity structure if someone creates a full declaration but then assigns to some function output? Or do we only allow the full sparsity declaration in the data and parameters blocks?

Reply:
The sparsity pattern has to be a known-at-compile-time deterministic function of the data.  Thankfully the Eigen sparse Cholesky (etc) does a two-phase sweep that can be separated out. Phase one is called a "symbolic factorization", which works out the sparsity pattern and does all the appropriate allocation. This is 100% a thing that can be used to infer that sparsity structure.

Some possible problems would be if someone wanted something like (apologies for the messy pseudocode)
```
sparse_matrix A; // structure given
sparse_matrix B; // structure given
sparse_matrix L; // structure to be inferred
real c; // data
L = chol(A + c*B);  // A problem if c = 0!
```

> That brings up another awkward point - do we need to check the sparsity structure if someone creates a full declaration but then assigns to some function output? Or do we only allow the full sparsity declaration in the data and parameters blocks?

Yeah. This is *the* issue here.  I had a shot at it here: https://aws1.discourse-cdn.com/standard14/uploads/mc_stan/original/2X/1/13fda4102c8f48e5aadbf0fbe75d3641a187d0a3.pdf

Essentially it would allow for two different declarations of sparse matrices: 
```
sparse_matrix[N, M, nz_foo_row_index, nz_foo_col_index] A;
sparse_matrix[foo(A)] L;
```
where `foo` is a function that returns a sparse matrix.  Steve's version doesn't have the explicit `[foo(A)]` bit and just has `sparse_matrix L = foo(A)` which is probably more in line with the idea of bringing declaration closer to use.  

This will work if we have a **requirement** that every function `foo_sparse` in math that returns a sparse matrix has a `foo_sparse_pattern_` variant that takes only data.  (This is the same pattern Eigen uses for sparse matrix factorizations.)  

@seantalts Would it be possible for the new compiler to find these things and just execute them once even though it would be in the model block? Because when this is declared in the parameter block, the pattern is data, while the "values" are `var`s   (Does this question make sense?)

A thing that would need to be avoided is 
```
// Declare B
sparse_matrix A = foo_sparse(B);
A = bar_sparse(B);
```
if `foo` and `bar` return different patterns. This means that the compound declaration has to be "special" (ie call the `foo_sparse_pattern_` function to declare the sparse matrix  and then call `foo_sparse` to get the values), while the ordinary `bar_sparse` call doesn't call `bar_sparse_pattern_`